core/src/main/java/io/grpc/CallCredentials2.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2016 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.concurrent.Executor;
+
+/**
+ * The new interface for {@link CallCredentials}.
+ *
+ * <p>THIS CLASS IS MEANT TO BE REFERENCED BY IMPLEMENTIONS ONLY.  All users should reference {@link
+ * CallCredentials}.
+ */
+public abstract class CallCredentials2 implements CallCredentials {
```

Comment:
This entire class should be marked experimental. It should deserve a note that it is part of a migration and is necessary to be used in the short-term but will be removed.

Reply:
I wanted to use the deprecation warnings on the old interface to push implementations to switch to the new interface. If we deprecate it now, implementations would have to suppress the deprecation warnings and may accidentally mask off implementations that have not been migrated, and would also cause confusion when we want implementations to switch from `CallCredentials2` back to `CallCredentials`. I think it's better to deprecate it after we replace `CallCredentials` with the new interface.

I have added the note, and an `@Internal` annotation to dissuade usage.

Reply:
I wasn't suggesting marking it deprecated, just experimental. I guess internal works as well, although it seems experimental is more appropriate since we do expect some "normal" users to use it. I'd normally prefer to keep internal a "if you are using this, you are probably doing it wrong."

Reply:
Sorry, read too fast :). Switched to experimental.